src/Lecture1.hs

```diff
@@ -1,126 +1,147 @@
+-- 2
+-- >>> lastDigit (-17)
+-- 7
+--
+-- 🕯 HINT: use the @mod@ function
+lastDigit :: Integer -> Integer
+lastDigit n = if n < 0 then mod (0 - n) 10 else mod n 10
+
+-- | Write a function that takes three numbers and returns the
+-- difference between the biggest number and the smallest one.
+--
+-- >>> minmax 7 1 4
+-- 6
+--
+-- Explanation: @minmax 7 1 4@ returns 6 because 7 is the biggest number
+-- and 1 is the smallest, and 7 - 1 = 6.
+--
+-- Try to use local variables (either let-in or where) to implement this
+-- function.
+minmax x y z = (tripletMax x y z) - (tripletMin x y z)
+  where
+    tripletMax x y z = max (max x y) (max y z)
+    tripletMin x y z = min (min x y) (min y z)
+
+-- | Implement a function that takes a string, start and end positions
+-- and returns a substring of a given string from the start position to
+-- the end (including).
+--
+-- >>> subString 3 7 "Hello, world!"
+-- "lo, w"
+--
+-- >>> subString 10 5 "Some very long String"
+-- ""
+--
+-- This function can accept negative start and end position. Negative
+-- start position can be considered as zero (e.g. substring from the
+-- first character) and negative end position should result in an empty
+-- string.
+subString :: Int -> Int -> [Char] -> [Char]
+subString start end str 
+    |end < 0 = ""
+    |start < 0 = take (end+1) str
+    |otherwise = take (end-start+1) (drop start str)
+
+-- | Write a function that takes a String — space separated numbers,
+-- and finds a sum of the numbers inside this string.
+--
+-- >>> strSum "100    -42  15"
+-- 73
+--
+-- The string contains only spaces and/or numbers.
+strSum str = sum (map strToInt (words str))
+  where
+    strToInt str = read str :: Integer
+
+-- | Write a function that takes a number and a list of numbers and
+-- returns a string, saying how many elements of the list are strictly
+-- greater than the given number and strictly lower.
+--
+-- >>> lowerAndGreater 3 [1 .. 9]
+-- "3 is greater than 2 elements and lower than 6 elements"
+--
+-- Explanation: the list [1 .. 9] contains 9 elements: [1, 2, 3, 4, 5, 6, 7, 8, 9]
+-- The given number 3 is greater than 2 elements (1 and 2)
+-- and lower than 6 elements (4, 5, 6, 7, 8 and 9).
+--
+-- 🕯 HINT: Use recursion to implement this function.
+lowerAndGreater :: Integer -> [Integer] -> [Char]
+lowerAndGreater n list =
+  show n
+    ++ " is greater than "
+    ++ show (countGreater n list 0)
+    ++ " elements and lower than "
+    ++ show (countLesser n list 0)
+    ++ " elements"
+  where
+    len = length list
+
+    countGreater :: Integer -> [Integer] -> Int -> Integer
+    countGreater num arr i
+      | i == len = 0
+      | num > (arr !! i) = 1 + countGreater n arr (i + 1)
+      | otherwise = countGreater num arr (i + 1)
+
+    countLesser :: Integer -> [Integer] -> Int -> Integer
+    countLesser num arr i
+      | i == len = 0
+      | num < (arr !! i) = 1 + countLesser n arr (i + 1)
+      | otherwise = countLesser num arr (i + 1)
```

Comment:
That's correct! 💯 

However, with this solution, you will walk through the list twice.
You can write it more performant with only one traversal of the list 🙂
It's desirable to write functions that traverse lists only a single time and consume elements only once. In that case, the composition of such functions will have nice streaming and lazy properties.